code/game/objects/items/RCD.dm

```diff
@@ -61,47 +64,69 @@ RLD
 	silo_mats = null
 	return ..()
 
+/obj/item/construction/melee_attack_chain(mob/user, atom/target, params)
+	if(istype(target, /obj/item/rcd_upgrade))
+		install_upgrade(target, user)
+		return
+	if(insert_matter(target, user))
+		return
+	return ..()
+
 /obj/item/construction/attackby(obj/item/W, mob/user, params)
-	if(iscyborg(user))
+	if(istype(W, /obj/item/rcd_upgrade))
+		install_upgrade(W, user)
+		return
+	if(insert_matter(W, user))
 		return
-	var/loaded = 0
-	if(istype(W, /obj/item/rcd_ammo))
-		var/obj/item/rcd_ammo/R = W
+	return ..()
+
+/// Installs an upgrade into the RCD checking if it is already installed, or if it is a banned upgrade
+/obj/item/construction/proc/install_upgrade(obj/item/rcd_upgrade/rcd_up, mob/user)
+	if(rcd_up.upgrade & upgrade)
+		to_chat(user, "<span class='warning'>[src] has already installed this upgrade!</span>")
+		return
+	if(rcd_up.upgrade & banned_upgrades)
+		to_chat(user, "<span class='warning'>[src] can't install this upgrade!</span>")
+		return
+	upgrade |= rcd_up.upgrade
+	if((rcd_up.upgrade & RCD_UPGRADE_SILO_LINK) && !silo_mats)
+		silo_mats = AddComponent(/datum/component/remote_materials, "RCD", FALSE, FALSE)
+	playsound(src.loc, 'sound/machines/click.ogg', 50, TRUE)
+	qdel(rcd_up)
+
+/// Inserts matter into the RCD allowing it to build
+/obj/item/construction/proc/insert_matter(obj/O, mob/user)
+	if(iscyborg(user))
+		return FALSE
+	var/loaded = FALSE
+	if(istype(O, /obj/item/rcd_ammo))
+		var/obj/item/rcd_ammo/R = O
 		var/load = min(R.ammoamt, max_matter - matter)
 		if(load <= 0)
 			to_chat(user, "<span class='warning'>[src] can't hold any more matter-units!</span>")
-			return
+			return FALSE
 		R.ammoamt -= load
 		if(R.ammoamt <= 0)
 			qdel(R)
 		matter += load
 		playsound(src.loc, 'sound/machines/click.ogg', 50, TRUE)
-		loaded = 1
-	else if(istype(W, /obj/item/stack/sheet/metal) || istype(W, /obj/item/stack/sheet/glass))
-		loaded = loadwithsheets(W, sheetmultiplier, user)
-	else if(istype(W, /obj/item/stack/sheet/plasteel))
-		loaded = loadwithsheets(W, plasteelmultiplier*sheetmultiplier, user) //12 matter for 1 plasteel sheet
-	else if(istype(W, /obj/item/stack/sheet/plasmarglass))
-		loaded = loadwithsheets(W, plasmarglassmultiplier*sheetmultiplier, user) //8 matter for one plasma rglass sheet
-	else if(istype(W, /obj/item/stack/sheet/rglass))
-		loaded = loadwithsheets(W, rglassmultiplier*sheetmultiplier, user) //6 matter for one rglass sheet
-	else if(istype(W, /obj/item/stack/rods))
-		loaded = loadwithsheets(W, sheetmultiplier * 0.5, user) // 2 matter for 1 rod, as 2 rods are produced from 1 metal
-	else if(istype(W, /obj/item/stack/tile/plasteel))
-		loaded = loadwithsheets(W, sheetmultiplier * 0.25, user) // 1 matter for 1 floortile, as 4 tiles are produced from 1 metal
+		loaded = TRUE
+	else if(istype(O, /obj/item/stack/sheet/metal) || istype(O, /obj/item/stack/sheet/glass))
```

Comment:
Feel free to ignore this request if you are not up to it, but this would work much better as a redefined proc.
Something like: `loaded = stack.load_with_sheets(src, user)`